File: src/p115/Solution.java
public class Solution {
	public int numDistinct(String S, String T) {
		int n = S.length(), m = T.length();
		int dp[][] = new int[n][];
		for (int i = 0; i < n; i++)
			dp[i] = new int[m];

		for (int i = 0; i < n; i++)
			if (S.charAt(i) == T.charAt(0))
				dp[i][0] = 1;

		for (int i = 1; i < m; i++) {
			int count = 0;
			for (int j = 0; j < n; j++) {
				if (S.charAt(j) == T.charAt(i))
					dp[j][i] += count;
				if (S.charAt(j) == T.charAt(i - 1))
					count += dp[j][i - 1];
			}
		}

		int res = 0;
		for (int i = 0; i < n; i++)
			res += dp[i][m - 1];
		return res;
	}

	public static void main(String[] args) {
		System.out.println(new Solution().numDistinct("rabbbit", "rabbit"));
	}
}

File: src/p115/solution.py
class Solution(object):
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        n, m = len(s), len(t)
        if n == 0:
            return 0
        dp = [[1] + [0] * m for _ in range(n + 1)]
        for i in range(n):
            for j in range(m):
                if s[i] == t[j]:
                    dp[i + 1][j + 1] = dp[i][j + 1] + dp[i][j]
                else:
                    dp[i + 1][j + 1] = dp[i][j + 1]
        return dp[n][m]
